src/Client/MultiplexedConnections.cpp

```diff
@@ -353,10 +359,25 @@ MultiplexedConnections::ReplicaState & MultiplexedConnections::getReplicaForRead
                 read_list.push_back(*connection->socket);
         }
 
-        int n = Poco::Net::Socket::select(read_list, write_list, except_list, settings.receive_timeout);
+        int n = Poco::Net::Socket::select(
+            read_list,
+            write_list,
+            except_list,
+            draining_timeout_seconds ? decltype(settings.receive_timeout)(draining_timeout_seconds) : settings.receive_timeout);
 
         if (n == 0)
+        {
+            for (ReplicaState & state : replica_states)
+            {
+                Connection * connection = state.connection;
+                if (connection != nullptr)
+                {
+                    connection->disconnect();
+                    invalidateReplica(state);
+                }
+            }
```

Comment:
Why this is required? Was it possible that after below exception the same connection can be used to send new queries and the protocol messages overlaps?

Reply:
> Was it possible that after below exception the same connection can be used to send new queries and the protocol messages overlaps?

Yes. The exception is likely related to timeout while receiving, and we cannot reuse this connection.

Reply:
Well, usually, in case of exception we disconnect anyway in `~RemoteQueryExecutor`. But I am ok to do it here too.
But could you add a comment?